### PR TITLE
crawler: solidify github process

### PR DESCRIPTION
**Pull request by flepied**
- manage errors from github
- run a thread per Github token
- dump raw pr into the /var/lib/crawler if it exists
- manage empty authors as ghost
- retry multiple times if a graphql request fails

### CODE REVIEW

File: monocle/crawler.py [hunk 1]
```diff
@@ -53,31 +58,90 @@ def __init__(self, args, elastic_conn='localhost:9200', elastic_timeout=10):
...
+                    )
+                    self.db.update(objects)
+                break
+            except pullrequest.ExtractPRIssue as err:
+                log.exception('Unable to decode PR')
+                log.error('idx=%d pr=%s' % (err.idx, err.pr))
+                # stop if we were on the last pr
+                if len(prs) == err.idx + 1:
+                    break
+                prs = prs[: err.idx] + prs[err.idx + 1 :]  # noqa
```
Comment (morucci) [inline monocle/crawler.py:116]: I don't get why we need to dump everything line 99.
Why to process extract_objects prs by bulk. It seems easier to me to skip and dump a failed extract in each gerrit and github crawler code directly. There is already the code to skip in each crawler so we can pass a dump_on_failed function to extract_objects.

File: monocle/github/pullrequest.py [hunk 2]
```diff
@@ -258,6 +268,11 @@ def get_one(self, org, repository, number):
         return (raw, self.extract_objects([raw]))
 
     def extract_objects(self, prs):
+        def get_login(data):
+            if data and 'login' in data and data['login']:
+                return data['login']
+            return 'ghost'
```
Comment (morucci) [inline monocle/github/pullrequest.py:274]: It is fine to have author at None.

Comment (flepied) [reply monocle/github/pullrequest.py:274]: this is the github account that is used for deleted users

File: monocle/crawler.py (hunk 1 above)
Comment (flepied) [reply monocle/crawler.py:116]: yes changing that

File: monocle/github/pullrequest.py [hunk 3]
```diff
@@ -386,24 +405,44 @@ def extract_pr_objects(pr):
                     'created_at': _commit.get('pushedDate', change['created_at']),
                 }
                 if _commit['author'].get('user'):
-                    obj['author'] = _commit['author']['user'].get('login')
+                    obj['author'] = get_login(_commit['author']['user'])
                 insert_change_attributes(obj, change)
                 objects.append(obj)
             return objects
 
+        def dump_data(data, prefix=None):
```
Comment (morucci) [inline monocle/github/pullrequest.py:413]: Do you think that's possible to have the same dump support for the Gerrit crawler ?

Comment (flepied) [reply monocle/github/pullrequest.py:413]: sure but let's do that in another PR

Comment (morucci) [reply monocle/github/pullrequest.py:413]: Ok. So let's merge this PR. I'm adding an issue to not forget about adding the same support for the Gerrit crawler.

Comment (flepied) [reply monocle/github/pullrequest.py]: done

**Pull request merged**